<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Introduction to R - part 2</title>
    <meta charset="utf-8" />
    <meta name="author" content="Tiago Nardi" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link rel="stylesheet" href="extra.css" type="text/css" />
    <link rel="stylesheet" href="xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Introduction to R - part 2
## üßâ
### Tiago Nardi
### University of Pavia

---





## Vectors


Let‚Äôs first create a __vector__, with 5 __elements__:

```r
x &lt;- c(2, 5, 8, 1, 2)
```
You can access elements inside a vector using their index position, using the square brackets *[ ]*


```r
x[2] # show the second element
```

```
## [1] 5
```
To select multiple elements use the concatenate function

```r
x[c(1,2)] # show the first and second element
```

```
## [1] 2 5
```
---
class: middle
## Indexing the vectors
Try these operations on your vector

```r
vector[3:5]
vector[c(1,3:5)]
vector[-2]
vector[vector &gt; 2]
```
Can you guess what you were extracting with these commands?
---
## Indexing the vectors
You can select a slice of the vector using the colon *:*

```r
x[3:5] # show all elements from a starting index (3) to an ending index (5) (slicing)
```

```
## [1] 8 1 2
```
You can combine both methods

```r
x[c(1,3:5)]
```

```
## [1] 2 8 1 2
```
You can get all the elements barring the elements with a minus before it *-*

```r
x[-2]
```

```
## [1] 2 8 1 2
```
---
## Indexing the vectors
Chose a numerical value and try these commands with you vector

```r
vector[vector &gt; value]
vector &gt; value
```
Why these results are different?

Can you guess it?

Write also a vector with only boolean values (TRUE/FALSE) with the same length of your first vector

```r
vector2 &lt;- c(FALSE,TRUE,TRUE,FALSE,FALSE)
vector[vector2]
```
---
## Indexing the vectors
You can select elements in the vector using logical conditions

```r
x[x &gt; 2] # show all elements greater than 2
```

```
## [1] 5 8
```
To access the vector you have to specify the values *inside* the brackets.

Otherwise you apply the logical operation to all elements in the vector

```r
x &gt; 2
```

```
## [1] FALSE  TRUE  TRUE FALSE FALSE
```

You can supply a logical vector to get only the elements in the positions with TRUE

```r
mr_boole &lt;- c(FALSE,TRUE,TRUE,FALSE,FALSE)
x[mr_boole]
```

```
## [1] 5 8
```
---
## Vectors
As you can assign values to an __object__

You can assign values inside a vector


```r
x[3]
```

```
## [1] 8
```

```r
x[3] &lt;- 11 # replace the third element with the number 11
x[3]
```

```
## [1] 11
```

---
## Rows and columns
Let's load the usual table (remember the working directory)

```r
db &lt;- read.csv("patric_redux.csv", header = T)
db1 &lt;- db # let's make a copy of the data frame
head(db)
```

```
##           Species    ID Contigs Genome.Length GC.Content PATRIC.CDS
## 1 Yersinia pestis AS539     250       4572127       47.5       4398
## 2 Yersinia pestis AS147     277       4592682       47.5       4485
## 3 Yersinia pestis AS134     237       4572981       47.5       4378
## 4 Yersinia pestis AS509     263       4605070       47.5       4378
## 5 Yersinia pestis  A251     201       4593919       47.5       4507
## 6 Yersinia pestis   24H     140       4492822       47.6       4620
##   Isolation_location Source
## 1           Isengard Hobbit
## 2          The Shire Hobbit
## 3              Rohan  Human
## 4             Mordor  Human
## 5          The Shire Hobbit
## 6             Erebor Dragon
```
---
## Rows and columns

To change the name use the *colnames()* functions

To get a column use the column number like a vector index

```r
colnames(db1) # It shows column names
```

```
## [1] "Species"            "ID"                 "Contigs"           
## [4] "Genome.Length"      "GC.Content"         "PATRIC.CDS"        
## [7] "Isolation_location" "Source"
```

```r
colnames(db1)[4] &lt;- "bp" # change the fourth column name
colnames(db1)
```

```
## [1] "Species"            "ID"                 "Contigs"           
## [4] "bp"                 "GC.Content"         "PATRIC.CDS"        
## [7] "Isolation_location" "Source"
```

```r
colnames(db1)[4] # Get only the fourth column
```

```
## [1] "bp"
```

---
## Rows and columns
How to set row names:

```r
rownames(db1) # It shows row names
```

```
##  [1] "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10" "11" "12" "13" "14" "15"
## [16] "16" "17" "18" "19" "20" "21" "22" "23" "24" "25" "26" "27" "28" "29" "30"
## [31] "31" "32" "33" "34" "35" "36" "37" "38" "39" "40" "41" "42" "43" "44" "45"
## [46] "46" "47" "48" "49" "50" "51" "52" "53" "54" "55" "56" "57" "58"
```

```r
rownames(db1) &lt;- db1$ID # We set as rownames all values from column
```

---
## Rows and columns


```r
head(db) # read the first 6 rows
```

```
##           Species    ID Contigs Genome.Length GC.Content PATRIC.CDS
## 1 Yersinia pestis AS539     250       4572127       47.5       4398
## 2 Yersinia pestis AS147     277       4592682       47.5       4485
## 3 Yersinia pestis AS134     237       4572981       47.5       4378
## 4 Yersinia pestis AS509     263       4605070       47.5       4378
## 5 Yersinia pestis  A251     201       4593919       47.5       4507
## 6 Yersinia pestis   24H     140       4492822       47.6       4620
##   Isolation_location Source
## 1           Isengard Hobbit
## 2          The Shire Hobbit
## 3              Rohan  Human
## 4             Mordor  Human
## 5          The Shire Hobbit
## 6             Erebor Dragon
```
---
## Rows and columns

```r
head(db1) # see the differences between this and the previous table
```

```
##               Species    ID Contigs      bp GC.Content PATRIC.CDS
## AS539 Yersinia pestis AS539     250 4572127       47.5       4398
## AS147 Yersinia pestis AS147     277 4592682       47.5       4485
## AS134 Yersinia pestis AS134     237 4572981       47.5       4378
## AS509 Yersinia pestis AS509     263 4605070       47.5       4378
## A251  Yersinia pestis  A251     201 4593919       47.5       4507
## 24H   Yersinia pestis   24H     140 4492822       47.6       4620
##       Isolation_location Source
## AS539           Isengard Hobbit
## AS147          The Shire Hobbit
## AS134              Rohan  Human
## AS509             Mordor  Human
## A251           The Shire Hobbit
## 24H               Erebor Dragon
```
---
## Cleaning up
To remove a column assign a __NULL__ value to it

```r
db1$Contigs &lt;- NULL # delete the column
```

If you want to remove an __object__ from R use

```r
rm(x)
```
It will free your memory from that specified __object__

The __object__ will disappear from the __environment tab__
---
## Accessing the data frame 
Extract columns and rows from a data frame (almost) like you access a __vector__

You can use the position indexes inside square brackets

First value is the row, the second the column

You can use position or names if you already set up row/column names

```r
dataframe_name[1, 1]
dataframe_name["row_name", "column_name"]
```
If don't specify one of the two values you get all elements (All rows/All columns)

```r
dataframe_name[,1] # all rows, first column
dataframe_name[1,] # first row, all column
```
---
## Accessing the data frame 

```r
dataframe_name["row_name", "column_name"]
dataframe_name[row_index, column_index]
dataframe_name[ , "column_name"]
dataframe_name[ , column_index]
dataframe_name["row_name", ]
dataframe_name[row_index,]
```
Try to access: 

the full row of sample *24H*

the *Source* of sample *EBD10-058*

the full column of *Isolation_location*

From the third to the 13th row the column *PATRIC.CDS"

All rows, excluding from the 4th row to the 10th, columns *bp* *GC.Content* 

---
## Accessing the data frame 

```r
db1["24H",]
```

```
##             Species  ID      bp GC.Content PATRIC.CDS Isolation_location Source
## 24H Yersinia pestis 24H 4492822       47.6       4620             Erebor Dragon
```

```r
db1["EBD10-058","Source"]
```

```
## [1] "Hobbit"
```

```r
db1[,"Isolation_location"]
```

```
##  [1] "Isengard"   "The Shire"  "Rohan"      "Mordor"     "The Shire" 
##  [6] "Erebor"     "Barad-dur"  "Mordor"     "Rohan"      "Rohan"     
## [11] "Lothlorien" "Rohan"      "Harad"      "Rohan"      "Gondor"    
## [16] "Gondor"     "Gondor"     "Gondor"     "Rohan"      "Gondor"    
## [21] "Gondor"     "The Shire"  "Gondor"     "Gondor"     "The Shire" 
## [26] "Harad"      "Rohan"      "Barad-dur"  "Barad-dur"  "Harad"     
## [31] "Gondor"     "The Shire"  "Lothlorien" "Lothlorien" "Erebor"    
## [36] "Erebor"     "Rivendell"  "Rivendell"  "Isengard"   "Lothlorien"
## [41] "Lothlorien" "Erebor"     "The Shire"  "Mirkwood"   "Gondor"    
## [46] "The Shire"  "Lothlorien" "Mordor"     "The Shire"  "Rivendell" 
## [51] "The Shire"  "Harad"      "Rohan"      "The Shire"  "Gondor"    
## [56] "The Shire"  "The Shire"  "The Shire"
```
---
## Accessing the data frame 

```r
db1[1:13,"PATRIC.CDS"] #check the rows inspecting db1
```

```
##  [1] 4398 4485 4378 4378 4507 4620 4364 4570 4555 4605 4806 4720 4566
```

```r
db1[-(4:10),c("bp","GC.Content")]
```

```
##                   bp GC.Content
## AS539        4572127   47.50000
## AS147        4592682   47.50000
## AS134        4572981   47.50000
## M70          4889404   47.09967
## M71          4745596   47.10000
## M72          4743182   47.23000
## M74          4744232   47.16000
## M32          4692341   47.20000
## FORC_017     4723612   46.80000
## FORC_013     4749283   47.30000
## FORC_004     4695233   47.22671
## FDAARGOS_227 5073657   47.47000
## FORC_002     3924069   42.59000
## NCTC10460    4767810   46.91510
## NCTC10462    3840239   46.80498
## LC18E-3      4676549   47.14303
## LC18E-9-14   4675119   47.14470
## M34          4982633   47.06526
## HR88         4745632   47.01089
## M29A27       5059344   47.09797
## PS23         4639377   47.08801
## FORC_003     4837317   47.04000
## ERL072344    4458462   46.92000
## IP01119      4643207   46.72000
## IP20322      4510028   46.98000
## IP05342      4722666   46.77000
## IP06077      4594630   46.79000
## M53          4744530   46.90000
## IP00178      4609560   46.84000
## 1116-D       4470393   47.56000
## YP173        4569957   47.53000
## YP261        4579841   47.52840
## YP139        3963071   48.33264
## YP1313       4598307   47.53702
## YP1906       4573713   47.52974
## YP2751-55    4582594   47.53343
## YP262        4532410   47.46285
## YP1240       4582524   47.53033
## YP2513       4537723   47.45978
## YP2514       4629574   47.50026
## YP111        4582792   47.53611
## PY-19        4510880   47.50000
## PY-76        4498987   47.50000
## PY-102       4672350   47.50000
## PY-103       4591865   47.80000
## PY-101       4591745   47.22671
## PY-22        4540396   47.60000
## PY-23        4537614   47.50000
## PY-24        4640238   47.70000
## PY-25        4540928   47.50000
## PY-77        4497645   47.70000
```
---
## Accessing the data frame 

```r
dataframe_name[ , c("column_1_name", "column_2_name")]
dataframe_name[ , c(first_column_index, second_column_index)]
dataframe_name[c("row_1_name", "row_2_name"), ]
dataframe_name[c(first_row_index, second_row_index), ]
dataframe_name[ , c(first_column_index : last_column_index)]
dataframe_name[c(first_row_index : last_row_index),]
```
You can mix and match different ways to access the data

```r
db[1,4] # First rown and 4th column
db[1,"Contigs"] #First row and column "Contigs"
db[35:78,] # From row 35 to 78 and all columns
db[,c(4,5)] #All rows and the 4th and 5th column
```
---
class: inverse, middle
.pull-left[
## Remember the typeof() function?
]
.pull-right[![](img/ein2.gif)]
---
## Object type
With the various form of extraction you get similar results but different data types

```r
dataframe_name[column_index]
dataframe_name[[column_index]] 
dataframe_name[, column_index] # Pay attention to the comma
dataframe_name$column_index
dataframe_name$column_index[row_index]
```
Chose a column from the database and try the diffferent options

Assign these values to a new __objects__

Look at how they appear in the environment

What's the difference?
---
## Object type

```r
typeof(db1[1])
```

```
## [1] "list"
```

```r
typeof(db1[, 1])
```

```
## [1] "character"
```

```r
typeof(db1[[1]])
```

```
## [1] "character"
```

```r
db1[1]
```

```
##                              Species
## AS539                Yersinia pestis
## AS147                Yersinia pestis
## AS134                Yersinia pestis
## AS509                Yersinia pestis
## A251                 Yersinia pestis
## 24H                  Yersinia pestis
## EBD10-058            Yersinia pestis
## M73            Yersinia kristensenii
## M47            Yersinia kristensenii
## HR100          Yersinia kristensenii
## M70            Yersinia kristensenii
## M71            Yersinia kristensenii
## M72            Yersinia kristensenii
## M74            Yersinia kristensenii
## M32          Yersinia enterocolitica
## FORC_017     Yersinia enterocolitica
## FORC_013     Yersinia enterocolitica
## FORC_004     Yersinia enterocolitica
## FDAARGOS_227 Yersinia enterocolitica
## FORC_002     Yersinia enterocolitica
## NCTC10460    Yersinia enterocolitica
## NCTC10462    Yersinia enterocolitica
## LC18E-3      Yersinia enterocolitica
## LC18E-9-14   Yersinia enterocolitica
## M34          Yersinia enterocolitica
## HR88         Yersinia enterocolitica
## M29A27       Yersinia enterocolitica
## PS23         Yersinia enterocolitica
## FORC_003     Yersinia enterocolitica
## ERL072344    Yersinia enterocolitica
## IP01119      Yersinia enterocolitica
## IP20322      Yersinia enterocolitica
## IP05342      Yersinia enterocolitica
## IP06077      Yersinia enterocolitica
## M53            Yersinia kristensenii
## IP00178      Yersinia enterocolitica
## 1116-D               Yersinia pestis
## YP173                Yersinia pestis
## YP261                Yersinia pestis
## YP139                Yersinia pestis
## YP1313               Yersinia pestis
## YP1906               Yersinia pestis
## YP2751-55            Yersinia pestis
## YP262                Yersinia pestis
## YP1240               Yersinia pestis
## YP2513               Yersinia pestis
## YP2514               Yersinia pestis
## YP111                Yersinia pestis
## PY-19                Yersinia pestis
## PY-76                Yersinia pestis
## PY-102               Yersinia pestis
## PY-103               Yersinia pestis
## PY-101               Yersinia pestis
## PY-22                Yersinia pestis
## PY-23                Yersinia pestis
## PY-24                Yersinia pestis
## PY-25                Yersinia pestis
## PY-77                Yersinia pestis
```
---
## Not Available
In R, missing values are represented by the symbol NA (not available)

To test for missing values

```r
y &lt;- c(1,2,3,NA)
is.na (y) #find which value is missing
```

```
## [1] FALSE FALSE FALSE  TRUE
```
To remove missing values from a data frame:

```r
db2 &lt;- db1[complete.cases(db1), ] # remove all rows with NA values
```
---
class: middle
## Data frame functions
*Table* creates a tabular results of categorical variables

```r
table(datafra$column) 
```

```r
table(db1$Isolation_location)
```

```
## 
##  Barad-dur     Erebor     Gondor      Harad   Isengard Lothlorien   Mirkwood 
##          3          4         11          4          2          6          1 
##     Mordor  Rivendell      Rohan  The Shire 
##          3          3          8         13
```

---
class: middle
## Data frame functions
*gsub* takes an input value and replace it with another one

```r
db1$Isolation_location&lt;-gsub("Barad-dur",
                             "Varallo Pombia",
                             db1$Isolation_location)
```
the word ‚ÄúBarad-dur‚Äù is replaced with ‚ÄúVarallo Pombia‚Äù and save the replaced column in the dataframe

Try to replace something from the column *Source*
---
class: middle
## Data frame functions
*unique* identify and eliminate duplicated values in an object. The object can be a vector or
a data frame. On data frames the function removes duplicated rows

```r
country &lt;- db1$Isolation_location
unique_country &lt;- unique(country)
```
---
class: middle
## Functions on data frames
&lt;img src="img/cowboy-bebop-cowboy.gif" style="width: 30%"/&gt;

How to check if a value is present in a column or vector:

```r
value %in% vector
```
To extract the first location that has the value of interest:

```r
match("value", dataframe_name$column_name)
```
Use these functions to see if one value of you choice is present in the column *Isolation_location*, and see which row has the first occurrence 
---
class: middle
## Functions on data frames

```r
"Varallo Pombia" %in% db1$Isolation_location
```

```
## [1] TRUE
```


```r
match("Varallo Pombia", db1$Isolation_location)
```

```
## [1] 7
```

```r
db1$Isolation_location[7]
```

```
## [1] "Varallo Pombia"
```
---
class: middle, inverse
.pull-left[## Subsetting and Binding]
.pull-right[

### Cutting, pasting and modifying data frames ![](img/cut.gif)]
---
## Subsetting
From a table, extract only a part of it, using a __pattern__


```r
subset(dataframe_name, dataframe_name$column_name logical_condition)
```
For example, if we want to keep genomes not too fragmented we can select elements with less than 100 contigs

```r
decent_qual &lt;- subset(db, db$Contigs &lt; 100)
```
You can use more than one logical condition using AND(*&amp;*) OR (*|*), or just using two subsets

```r
gondor &lt;- subset(db,db$Contigs&lt;100 &amp;
                 db$Isolation_location=="Gondor")
```
You can subset for columns:

```r
red_table &lt;- subset(db,select = c(1, 3, 4, 5))
```
---
# Logical operators

<div id="jxsxpzzzmk" style="overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>html {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
}

#jxsxpzzzmk .gt_table {
  display: table;
  border-collapse: collapse;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#jxsxpzzzmk .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#jxsxpzzzmk .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#jxsxpzzzmk .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 0;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#jxsxpzzzmk .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#jxsxpzzzmk .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#jxsxpzzzmk .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#jxsxpzzzmk .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#jxsxpzzzmk .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#jxsxpzzzmk .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#jxsxpzzzmk .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#jxsxpzzzmk .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
}

#jxsxpzzzmk .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#jxsxpzzzmk .gt_from_md > :first-child {
  margin-top: 0;
}

#jxsxpzzzmk .gt_from_md > :last-child {
  margin-bottom: 0;
}

#jxsxpzzzmk .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#jxsxpzzzmk .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#jxsxpzzzmk .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#jxsxpzzzmk .gt_row_group_first td {
  border-top-width: 2px;
}

#jxsxpzzzmk .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#jxsxpzzzmk .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#jxsxpzzzmk .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#jxsxpzzzmk .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#jxsxpzzzmk .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#jxsxpzzzmk .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#jxsxpzzzmk .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#jxsxpzzzmk .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#jxsxpzzzmk .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#jxsxpzzzmk .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-left: 4px;
  padding-right: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#jxsxpzzzmk .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#jxsxpzzzmk .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#jxsxpzzzmk .gt_left {
  text-align: left;
}

#jxsxpzzzmk .gt_center {
  text-align: center;
}

#jxsxpzzzmk .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#jxsxpzzzmk .gt_font_normal {
  font-weight: normal;
}

#jxsxpzzzmk .gt_font_bold {
  font-weight: bold;
}

#jxsxpzzzmk .gt_font_italic {
  font-style: italic;
}

#jxsxpzzzmk .gt_super {
  font-size: 65%;
}

#jxsxpzzzmk .gt_footnote_marks {
  font-style: italic;
  font-weight: normal;
  font-size: 75%;
  vertical-align: 0.4em;
}

#jxsxpzzzmk .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#jxsxpzzzmk .gt_slash_mark {
  font-size: 0.7em;
  line-height: 0.7em;
  vertical-align: 0.15em;
}

#jxsxpzzzmk .gt_fraction_numerator {
  font-size: 0.6em;
  line-height: 0.6em;
  vertical-align: 0.45em;
}

#jxsxpzzzmk .gt_fraction_denominator {
  font-size: 0.6em;
  line-height: 0.6em;
  vertical-align: -0.05em;
}
</style>
<table class="gt_table">
  
  <thead class="gt_col_headings">
    <tr>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">Operator</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">Description</th>
    </tr>
  </thead>
  <tbody class="gt_table_body">
    <tr><td class="gt_row gt_left">&gt;</td>
<td class="gt_row gt_left">Greater than</td></tr>
    <tr><td class="gt_row gt_left">&gt;=</td>
<td class="gt_row gt_left">Greater than or equal tp</td></tr>
    <tr><td class="gt_row gt_left">&lt;</td>
<td class="gt_row gt_left">Lesser than</td></tr>
    <tr><td class="gt_row gt_left">&lt;=</td>
<td class="gt_row gt_left">Lesser than or equal to</td></tr>
    <tr><td class="gt_row gt_left">==</td>
<td class="gt_row gt_left">Equal to</td></tr>
    <tr><td class="gt_row gt_left">!=</td>
<td class="gt_row gt_left">Not equal to</td></tr>
    <tr><td class="gt_row gt_left">&amp;</td>
<td class="gt_row gt_left">AND</td></tr>
    <tr><td class="gt_row gt_left">|</td>
<td class="gt_row gt_left">OR</td></tr>
    <tr><td class="gt_row gt_left">!</td>
<td class="gt_row gt_left">NOT</td></tr>
  </tbody>
  
  
</table>
</div>
---
class: middle
## Exercise
&lt;img src="img/cowboy-bebop-cowboy.gif" style="width: 30%"/&gt;

Get a table with only genomes from Elf and 4500000 as genome length

```r
subset(dataframe_name, dataframe_name$column_name logical_condition)
```
---
## Exercise

```r
rohan &lt;- subset(db,db$Genome.Length &gt; 4500000 &amp;
                 db$Source=="Elf")
```
<div id="bolzrqynss" style="overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>html {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
}

#bolzrqynss .gt_table {
  display: table;
  border-collapse: collapse;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#bolzrqynss .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#bolzrqynss .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#bolzrqynss .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 0;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#bolzrqynss .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#bolzrqynss .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#bolzrqynss .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#bolzrqynss .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#bolzrqynss .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#bolzrqynss .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#bolzrqynss .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#bolzrqynss .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
}

#bolzrqynss .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#bolzrqynss .gt_from_md > :first-child {
  margin-top: 0;
}

#bolzrqynss .gt_from_md > :last-child {
  margin-bottom: 0;
}

#bolzrqynss .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#bolzrqynss .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#bolzrqynss .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#bolzrqynss .gt_row_group_first td {
  border-top-width: 2px;
}

#bolzrqynss .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#bolzrqynss .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#bolzrqynss .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#bolzrqynss .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#bolzrqynss .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#bolzrqynss .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#bolzrqynss .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#bolzrqynss .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#bolzrqynss .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#bolzrqynss .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-left: 4px;
  padding-right: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#bolzrqynss .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#bolzrqynss .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#bolzrqynss .gt_left {
  text-align: left;
}

#bolzrqynss .gt_center {
  text-align: center;
}

#bolzrqynss .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#bolzrqynss .gt_font_normal {
  font-weight: normal;
}

#bolzrqynss .gt_font_bold {
  font-weight: bold;
}

#bolzrqynss .gt_font_italic {
  font-style: italic;
}

#bolzrqynss .gt_super {
  font-size: 65%;
}

#bolzrqynss .gt_footnote_marks {
  font-style: italic;
  font-weight: normal;
  font-size: 75%;
  vertical-align: 0.4em;
}

#bolzrqynss .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#bolzrqynss .gt_slash_mark {
  font-size: 0.7em;
  line-height: 0.7em;
  vertical-align: 0.15em;
}

#bolzrqynss .gt_fraction_numerator {
  font-size: 0.6em;
  line-height: 0.6em;
  vertical-align: 0.45em;
}

#bolzrqynss .gt_fraction_denominator {
  font-size: 0.6em;
  line-height: 0.6em;
  vertical-align: -0.05em;
}
</style>
<table class="gt_table">
  
  <thead class="gt_col_headings">
    <tr>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">Species</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">ID</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">Contigs</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">Genome.Length</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">GC.Content</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1">PATRIC.CDS</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">Isolation_location</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1">Source</th>
    </tr>
  </thead>
  <tbody class="gt_table_body">
    <tr><td class="gt_row gt_left">Yersinia kristensenii</td>
<td class="gt_row gt_left">M70</td>
<td class="gt_row gt_right">81</td>
<td class="gt_row gt_right">4889404</td>
<td class="gt_row gt_right">47.09967</td>
<td class="gt_row gt_right">4806</td>
<td class="gt_row gt_left">Lothlorien</td>
<td class="gt_row gt_left">Elf</td></tr>
    <tr><td class="gt_row gt_left">Yersinia enterocolitica</td>
<td class="gt_row gt_left">IP05342</td>
<td class="gt_row gt_right">169</td>
<td class="gt_row gt_right">4722666</td>
<td class="gt_row gt_right">46.77000</td>
<td class="gt_row gt_right">4935</td>
<td class="gt_row gt_left">Lothlorien</td>
<td class="gt_row gt_left">Elf</td></tr>
    <tr><td class="gt_row gt_left">Yersinia enterocolitica</td>
<td class="gt_row gt_left">IP06077</td>
<td class="gt_row gt_right">159</td>
<td class="gt_row gt_right">4594630</td>
<td class="gt_row gt_right">46.79000</td>
<td class="gt_row gt_right">4743</td>
<td class="gt_row gt_left">Lothlorien</td>
<td class="gt_row gt_left">Elf</td></tr>
    <tr><td class="gt_row gt_left">Yersinia pestis</td>
<td class="gt_row gt_left">YP173</td>
<td class="gt_row gt_right">3</td>
<td class="gt_row gt_right">4569957</td>
<td class="gt_row gt_right">47.53000</td>
<td class="gt_row gt_right">4644</td>
<td class="gt_row gt_left">Rivendell</td>
<td class="gt_row gt_left">Elf</td></tr>
    <tr><td class="gt_row gt_left">Yersinia pestis</td>
<td class="gt_row gt_left">YP1313</td>
<td class="gt_row gt_right">235</td>
<td class="gt_row gt_right">4598307</td>
<td class="gt_row gt_right">47.53702</td>
<td class="gt_row gt_right">5203</td>
<td class="gt_row gt_left">Lothlorien</td>
<td class="gt_row gt_left">Elf</td></tr>
    <tr><td class="gt_row gt_left">Yersinia pestis</td>
<td class="gt_row gt_left">YP2514</td>
<td class="gt_row gt_right">212</td>
<td class="gt_row gt_right">4629574</td>
<td class="gt_row gt_right">47.50026</td>
<td class="gt_row gt_right">4766</td>
<td class="gt_row gt_left">Lothlorien</td>
<td class="gt_row gt_left">Elf</td></tr>
  </tbody>
  
  
</table>
</div>
---
class: middle
## Dropping levels
When subsetting data frames there may be some levels (for factors) that are unused

Remove them because they may alter downstream analyses

```r
table(data_frame$column)
```
If you get a factor with a value of 0 you should  drop the levels

```r
dataframe_name &lt;- droplevels (dataframe_name)
```
---
class: middle
## Binding
.pull-left[
#### Combining vectors to a a dataframe:

- #### ___c___*bind* to bind a __c__olumn to a dataframe

- #### ___r___*bind* to bind a __r__ow to a dataframe

- #### vectors have to be of the same lenght]
.pull-right[![](img/binding.png)]
---
class: middle
# Binding - Pasting

Paste combine strings to a single string, or two vectors of strings

```r
new_vector&lt;-paste(string_vector1,string_vector2,sep=" ")
```
In the new string/vector the original strings will be separated by the separator (*sep=*)

If you don't want any separation write *sep=""*

You can assign the new vector to a dataframe or use it to replace a column

You can use a fixed string

```r
new_vector&lt;-paste("fixed_string",string_vector2,sep="")
```
---
class: middle
## Exercise
&lt;img src="img/cowboy-bebop-cowboy.gif" style="width: 30%"/&gt;

Create 2 subsets:
- one with only samples from *Erebor* 
- one with only samples from *Barad-dur*

Bind the two tables into a single dataframe, and add a column pasting *Isolation_location* and *Source* columns

```r
new_df &lt;- rbind(df1,df2) #new_r now is a new row of the data frame
new_vector&lt;-paste(string_vector1,string_vector2,sep=" ")
```
---
class: middle
# Exercise


```r
Erebor &lt;- subset(db, db$Isolation_location=="Erebor")
Barad_dur &lt;- subset(db, db$Isolation_location=="Barad-dur")
db2 &lt;- rbind(Erebor, Barad_dur)
db2$newco &lt;- paste(db2$Isolation_location,db2$Source)
```
---
# Sort
Sort a data frame in R (by default, sorting is ASCENDING)

```r
dataframe_name &lt;- dataframe_name[order(dataframe_name$column), ]
```

```
db2&lt;- db2[order(db2$Contigs),] # sort the whole data frame by column ‚ÄúContigs‚Äù
```
---
## Vector functions
These can be applied in the same way on vectors

```r
function(vector)
```
*length* returns the length of the vector (or the columns number in a dataframe)

*max* returns the maximum value of a numeric vector or column

*min* returns the minimum value of a numeric vector or column

*range* it outputs the smallest and largest values of the numeric vector or column
---
# Exercise
&lt;img src="img/cowboy-bebop-cowboy.gif" style="width: 30%"/&gt;

- Get the range of the *GC.Content*

- Get the max value of contigs in only Hobbit samples

- Get the minimum value of *PATRIC.CDS* in samples with less than 100 contigs
---
## Results

```r
range(db$GC.Content)
```

```
## [1] 42.59000 48.33264
```

```r
hobbit &lt;- subset(db, db$Source == "Hobbit")
max(hobbit$Contigs)
```

```
## [1] 581
```

```r
contig_100 &lt;- subset(db, db$Contigs &lt; 100)
min(contig_100$PATRIC.CDS)
```

```
## [1] 4124
```
---
class: inverse, middle
## Good things come in small packages
---
## Install and load libraries
R, by default, only comes with a restricted number of functions

To extend the functionalities we can install external packages 

To install a package in R:

```r
install.packages("library_name")
```
after you have installed you can load the package

```r
# There's a time and place for everything
# but do not run this code now
install.packages(‚Äúggplot2‚Äù)
library(ggplot2)
```
---
![](img/ikea1.png)
## install.packages("package")
---
![](img/ikea2.png)

## library(package)
---
class: inverse, middle, center
# EXTRA
---
class: middle
## Libraries

- Arguably __tidyverse__ is the most important set of packages, including:

  - *Dplyr*: makes data manipulation much more easier

  - *ggplot2*: extensive system to create graphics, extremely powerful
- *Bioconductor*: series of packages for biology and genomics
- *RColorBrewer*: useful tool to manage color palettes in R
- *Xaringan*: to use markdown to write slides (including these slide)
---
# Dplyr
dplyr is a package for data manipulation, providing a consistent set of functions that help
you solve the most common data manipulation challenges

```r
install.packages (‚Äúdplyr‚Äù)
library (dplyr)
```
A common use is the *join* function

In order to combine two data frames, at least one column must be present in both data
frames and have the same header


```r
full &lt;- full_join(db1,db2,by="Genome_ID") # all rows
inner &lt;- inner_join(db1,db2,by="Genome_ID") # rows present in both
left &lt;- left_join(db1,db2,by="Genome_ID") # all rows in db1
right &lt;- right_join(db1,db2,by="Genome_ID") # all rows in db2
```
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>
<style>
.logo {
  background-image: url(img/midi_trans.png);
  background-size: contain;
  background-repeat: no-repeat;
  position: absolute;
  top: 1em;
  right: 1em;
  width: 110px;
  height: 128px;
  z-index: 0;
}
</style>

<script>
document
  .querySelectorAll(
    '.remark-slide-content' +
    ':not(.title-slide)' +
    // add additional classes to exclude here, e.g.
    // ':not(.inverse)' +
    ':not(.hide-logo)'
  )
  .forEach(el => {
    el.innerHTML += '<div class="logo"></div>';
  });
</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
